File: backend/app/routers/notifications.py
from fastapi import APIRouter, Depends, HTTPException, status, WebSocket, WebSocketDisconnect
from sqlalchemy.orm import Session
from sqlalchemy import desc, and_, or_
from typing import List, Optional
from pydantic import BaseModel
from datetime import datetime, timezone
import uuid
import logging

from app.db.session import get_db
from app.models.user import User
from app.models.notification import Notification, PushSubscription
from app.models.system_config import SystemConfig
from app.schemas.system_config import SystemConfigResponse
from app.auth.dependencies import get_current_active_user, get_current_admin
from app.core.roles import Role
from app.core.config import settings
from app.services.push_notifications import send_web_push, send_web_push_to_multiple_users
from app.services.websocket_manager import manager

router = APIRouter(prefix="/notifications", tags=["notifications"])
logger = logging.getLogger(__name__)


# ===== SCHEMAS =====

class NotificationCreate(BaseModel):
    title: str
    message: str
    type: str = "info"  # info, success, warning, error
    link: Optional[str] = None


class NotificationResponse(BaseModel):
    id: str
    user_id: str
    title: str
    message: str
    type: str
    link: Optional[str]
    is_read: bool
    is_muted: bool
    sent_by: Optional[str]
    created_at: datetime
    read_at: Optional[datetime]

    class Config:
        from_attributes = True


class BulkNotificationCreate(BaseModel):
    title: str
    message: str
    type: str = "info"
    link: Optional[str] = None
    user_ids: Optional[List[str]] = None  # Specific users, or None for all
    role: Optional[str] = None  # Send to specific role: USER, ADMIN, SUPERUSER


class PushSubscriptionCreate(BaseModel):
    endpoint: str
    p256dh: str
    auth: str


class NotificationStats(BaseModel):
    total: int
    unread: int
    muted: int


# ===== USER ENDPOINTS =====

@router.get("/", response_model=List[NotificationResponse])
async def get_my_notifications(
    limit: int = 50,
    offset: int = 0,
    unread_only: bool = False,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get current user's notifications"""
    query = db.query(Notification).filter(Notification.user_id == current_user.id)

    if unread_only:
        query = query.filter(Notification.is_read == False)

    notifications = query.order_by(desc(Notification.created_at)).limit(limit).offset(offset).all()
    return notifications


@router.get("/stats", response_model=NotificationStats)
async def get_notification_stats(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get notification statistics for current user"""
    total = db.query(Notification).filter(Notification.user_id == current_user.id).count()
    unread = db.query(Notification).filter(
        and_(Notification.user_id == current_user.id, Notification.is_read == False)
    ).count()
    muted = db.query(Notification).filter(
        and_(Notification.user_id == current_user.id, Notification.is_muted == True)
    ).count()

    return {"total": total, "unread": unread, "muted": muted}


@router.patch("/{notification_id}/read")
async def mark_as_read(
    notification_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Mark a notification as read"""
    notification = db.query(Notification).filter(
        and_(Notification.id == notification_id, Notification.user_id == current_user.id)
    ).first()

    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")

    notification.is_read = True
    notification.read_at = datetime.now(timezone.utc)
    db.commit()

    return {"message": "Notification marked as read"}


@router.patch("/read-all")
async def mark_all_as_read(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Mark all notifications as read"""
    db.query(Notification).filter(
        and_(Notification.user_id == current_user.id, Notification.is_read == False)
    ).update({"is_read": True, "read_at": datetime.now(timezone.utc)})
    db.commit()

    return {"message": "All notifications marked as read"}


@router.patch("/{notification_id}/mute")
async def mute_notification(
    notification_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Mute a notification"""
    notification = db.query(Notification).filter(
        and_(Notification.id == notification_id, Notification.user_id == current_user.id)
    ).first()

    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")

    notification.is_muted = True
    db.commit()

    return {"message": "Notification muted"}


@router.delete("/{notification_id}")
async def delete_notification(
    notification_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Delete a notification"""
    notification = db.query(Notification).filter(
        and_(Notification.id == notification_id, Notification.user_id == current_user.id)
    ).first()

    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")

    db.delete(notification)
    db.commit()

    return {"message": "Notification deleted"}


# ===== WEBSOCKET ENDPOINT =====

@router.websocket("/ws/{user_id}")
async def websocket_endpoint(
    websocket: WebSocket,
    user_id: str,
    token: str
):
    """
    WebSocket endpoint for real-time notifications

    Usage: ws://localhost:8000/api/notifications/ws/{user_id}?token={jwt_token}
    """
    # Authenticate the user via token
    try:
        # We need to validate the token manually since WebSocket doesn't support Depends
        from jose import jwt, JWTError
        from app.auth.auth import SECRET_KEY, ALGORITHM

        try:
            payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
            token_user_id: str = payload.get("sub")

            if token_user_id is None or token_user_id != user_id:
                logger.warning(f"WebSocket auth failed: token user_id {token_user_id} != requested {user_id}")
                await websocket.close(code=1008, reason="Unauthorized")
                return
        except JWTError as e:
            logger.warning(f"WebSocket JWT validation failed: {e}")
            await websocket.close(code=1008, reason="Invalid token")
            return

    except Exception as e:
        logger.error(f"WebSocket authentication error: {e}")
        await websocket.close(code=1011, reason="Authentication failed")
        return

    # Connect the user
    await manager.connect(websocket, user_id)

    try:
        # Send initial connection confirmation
        await websocket.send_json({
            "type": "connected",
            "message": "WebSocket connection established"
        })

        # Keep connection alive and listen for messages
        while True:
            try:
                # Receive messages from client (ping/pong for keepalive)
                data = await websocket.receive_text()

                if data == "ping":
                    await websocket.send_json({"type": "pong"})

            except WebSocketDisconnect:
                break
            except Exception as e:
                logger.error(f"Error in WebSocket receive loop: {e}")
                break

    finally:
        manager.disconnect(websocket, user_id)


# ===== PUSH SUBSCRIPTION ENDPOINTS =====

@router.get("/vapid-public-key")
async def get_vapid_public_key():
    """Get VAPID public key for push notifications"""
    if not settings.VAPID_PUBLIC_KEY:
        raise HTTPException(status_code=503, detail="Push notifications are not configured")
    return {"publicKey": settings.VAPID_PUBLIC_KEY}


@router.post("/subscribe")
async def subscribe_to_push(
    subscription: PushSubscriptionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Subscribe to push notifications"""
    # Check if subscription already exists
    existing = db.query(PushSubscription).filter(
        PushSubscription.endpoint == subscription.endpoint
    ).first()

    if existing:
        # Update user_id if changed
        if existing.user_id != current_user.id:
            existing.user_id = current_user.id
        existing.last_used = datetime.now(timezone.utc)
        db.commit()
        return {"message": "Subscription updated", "id": existing.id}

    # Create new subscription
    new_subscription = PushSubscription(
        id=str(uuid.uuid4()),
        user_id=current_user.id,
        endpoint=subscription.endpoint,
        p256dh=subscription.p256dh,
        auth=subscription.auth
    )
    db.add(new_subscription)
    db.commit()

    return {"message": "Subscribed to push notifications", "id": new_subscription.id}


@router.delete("/unsubscribe")
async def unsubscribe_from_push(
    endpoint: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Unsubscribe from push notifications"""
    subscription = db.query(PushSubscription).filter(
        and_(PushSubscription.endpoint == endpoint, PushSubscription.user_id == current_user.id)
    ).first()

    if subscription:
        db.delete(subscription)
        db.commit()

    return {"message": "Unsubscribed from push notifications"}


# ===== ADMIN ENDPOINTS =====

@router.post("/send", status_code=status.HTTP_201_CREATED)
async def send_notification(
    notification: BulkNotificationCreate,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin)
):
    """Send notification to users (Admin/Superuser only)"""

    # Determine target users
    if notification.user_ids:
        # Send to specific users
        target_users = db.query(User).filter(User.id.in_(notification.user_ids)).all()
    elif notification.role:
        # Send to all users with specific role
        target_users = db.query(User).filter(User.role == notification.role).all()
    else:
        # Send to all users
        target_users = db.query(User).all()

    if not target_users:
        raise HTTPException(status_code=404, detail="No users found")

    # Create notifications for all target users
    notifications_created = []
    for user in target_users:
        new_notification = Notification(
            id=str(uuid.uuid4()),
            user_id=user.id,
            title=notification.title,
            message=notification.message,
            type=notification.type,
            link=notification.link,
            sent_by=admin.id
        )
        db.add(new_notification)
        notifications_created.append(new_notification.id)

    db.commit()

    # Send real-time WebSocket notifications to connected users
    websocket_sent = 0
    for user in target_users:
        try:
            await manager.send_notification_update(
                user_id=user.id,
                notification_id=notifications_created[target_users.index(user)],
                title=notification.title,
                message=notification.message,
                notification_type=notification.type,
                link=notification.link
            )
            websocket_sent += 1
        except Exception as e:
            logger.error(f"Error sending WebSocket notification to user {user.id}: {e}")

    # Send web push notifications to all target users
    user_ids = [user.id for user in target_users]
    push_result = send_web_push_to_multiple_users(
        db=db,
        user_ids=user_ids,
        title=notification.title,
        message=notification.message,
        notification_type=notification.type,
        link=notification.link
    )

    return {
        "message": f"Notification sent to {len(target_users)} users",
        "count": len(target_users),
        "notification_ids": notifications_created,
        "websocket_sent": websocket_sent,
        "push_notifications": push_result
    }


@router.get("/admin/all")
async def get_all_notifications(
    limit: int = 100,
    offset: int = 0,
    user_id: Optional[str] = None,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin)
):
    """Get all notifications (Admin/Superuser only)"""
    query = db.query(Notification)

    if user_id:
        query = query.filter(Notification.user_id == user_id)

    total = query.count()
    notifications = query.order_by(desc(Notification.created_at)).limit(limit).offset(offset).all()

    return {
        "notifications": notifications,
        "total": total,
        "limit": limit,
        "offset": offset
    }


@router.get("/admin/stats")
async def get_admin_notification_stats(
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin)
):
    """Get global notification statistics (Admin/Superuser only)"""
    total = db.query(Notification).count()
    total_unread = db.query(Notification).filter(Notification.is_read == False).count()
    total_muted = db.query(Notification).filter(Notification.is_muted == True).count()
    total_subscriptions = db.query(PushSubscription).count()

    return {
        "total_notifications": total,
        "total_unread": total_unread,
        "total_muted": total_muted,
        "total_push_subscriptions": total_subscriptions
    }


@router.get("/config", response_model=SystemConfigResponse)
async def get_notification_config(
    db: Session = Depends(get_db)
):
    """Get system-wide notification configuration"""
    config = db.query(SystemConfig).filter(SystemConfig.key == "notification_prompt_message").first()
    if not config:
        # Return default if not set in DB
        return {
            "key": "notification_prompt_message",
            "value": "To receive AI-generated messages and important updates, please allow notification access.",
            "updated_at": datetime.now(timezone.utc)
        }
    return config

File: services/notifications.ts
/**
 * Notification Service
 * Handles in-app and push notifications
 */

import { get, post, patch, del, API_BASE_URL, getWebSocketBaseUrl } from './api';
import { getToken } from './auth';

export interface Notification {
  id: string;
  user_id: string;
  title: string;
  message: string;
  type: 'info' | 'success' | 'warning' | 'error';
  link?: string;
  is_read: boolean;
  is_muted: boolean;
  sent_by?: string;
  created_at: string;
  read_at?: string;
}

export interface NotificationStats {
  total: number;
  unread: number;
  muted: number;
}

export interface BulkNotificationRequest {
  title: string;
  message: string;
  type?: 'info' | 'success' | 'warning' | 'error';
  link?: string;
  user_ids?: string[];
  role?: 'USER' | 'ADMIN' | 'SUPERUSER';
}

// ===== USER FUNCTIONS =====

/**
 * Get current user's notifications
 */
export async function getNotifications(
  limit: number = 50,
  offset: number = 0,
  unreadOnly: boolean = false
): Promise<Notification[]> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  const params = new URLSearchParams({
    limit: limit.toString(),
    offset: offset.toString(),
    unread_only: unreadOnly.toString(),
  });

  return get<Notification[]>(`/notifications?${params}`, token);
}

/**
 * Get notification statistics
 */
export async function getNotificationStats(): Promise<NotificationStats> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  return get<NotificationStats>('/notifications/stats', token);
}

/**
 * Mark a notification as read
 */
export async function markAsRead(notificationId: string): Promise<void> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  await patch(`/notifications/${notificationId}/read`, {}, token);
}

/**
 * Mark all notifications as read
 */
export async function markAllAsRead(): Promise<void> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  await patch('/notifications/read-all', {}, token);
}

/**
 * Mute a notification
 */
export async function muteNotification(notificationId: string): Promise<void> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  await patch(`/notifications/${notificationId}/mute`, {}, token);
}

/**
 * Delete a notification
 */
export async function deleteNotification(notificationId: string): Promise<void> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  try {
    await del(`/notifications/${notificationId}`, undefined, token);
    console.log(`Notification ${notificationId} deleted successfully`);
  } catch (error) {
    console.error('Failed to delete notification:', error);
    throw error;
  }
}

// ===== PUSH NOTIFICATION FUNCTIONS =====

/**
 * Request push notification permission
 */
export async function requestPushPermission(): Promise<NotificationPermission> {
  if (!('Notification' in window)) {
    console.warn('Push notifications not supported');
    return 'denied';
  }

  return Notification.requestPermission();
}

/**
 * Get VAPID public key from backend
 */
export async function getVapidPublicKey(): Promise<string> {
  try {
    const response = await get<{ publicKey: string }>('/notifications/vapid-public-key');
    return response.publicKey;
  } catch (error) {
    console.error('Failed to get VAPID public key:', error);
    throw error;
  }
}

/**
 * Subscribe to push notifications
 */
export async function subscribeToPush(): Promise<boolean> {
  try {
    const token = getToken();
    if (!token) throw new Error('Authentication required');

    // Request permission first
    const permission = await requestPushPermission();
    if (permission !== 'granted') {
      return false;
    }

    // Register service worker
    if (!('serviceWorker' in navigator)) {
      console.warn('Service workers not supported');
      return false;
    }

    const registration = await navigator.serviceWorker.register('/sw.js');
    await navigator.serviceWorker.ready;

    // Get VAPID public key from backend
    const vapidPublicKey = await getVapidPublicKey();

    // Get push subscription
    const subscription = await registration.pushManager.subscribe({
      userVisibleOnly: true,
      applicationServerKey: urlBase64ToUint8Array(vapidPublicKey) as BufferSource,
    });

    // Send subscription to backend
    const subscriptionJSON = subscription.toJSON();
    await post(
      '/notifications/subscribe',
      {
        endpoint: subscriptionJSON.endpoint,
        p256dh: subscriptionJSON.keys?.p256dh,
        auth: subscriptionJSON.keys?.auth,
      },
      token
    );

    return true;
  } catch (error) {
    console.error('Failed to subscribe to push notifications:', error);
    return false;
  }
}

/**
 * Unsubscribe from push notifications
 */
export async function unsubscribeFromPush(): Promise<boolean> {
  try {
    const token = getToken();
    if (!token) throw new Error('Authentication required');

    if (!('serviceWorker' in navigator)) return false;

    const registration = await navigator.serviceWorker.getRegistration();
    if (!registration) return false;

    const subscription = await registration.pushManager.getSubscription();
    if (!subscription) return false;

    // Unsubscribe from backend
    await del(
      `/notifications/unsubscribe?endpoint=${encodeURIComponent(subscription.endpoint)}`,
      undefined,
      token
    );

    // Unsubscribe from browser
    await subscription.unsubscribe();

    return true;
  } catch (error) {
    console.error('Failed to unsubscribe from push notifications:', error);
    return false;
  }
}

/**
 * Check if push notifications are enabled
 */
export async function isPushEnabled(): Promise<boolean> {
  try {
    if (!('serviceWorker' in navigator) || !('Notification' in window)) {
      return false;
    }

    if (Notification.permission !== 'granted') {
      return false;
    }

    const registration = await navigator.serviceWorker.getRegistration();
    if (!registration) return false;

    const subscription = await registration.pushManager.getSubscription();
    return subscription !== null;
  } catch {
    return false;
  }
}

// ===== ADMIN FUNCTIONS =====

/**
 * Send bulk notification (Admin/Superuser only)
 */
export async function sendBulkNotification(
  notification: BulkNotificationRequest
): Promise<{ message: string; count: number }> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  return post('/notifications/send', notification, token);
}

/**
 * Get all notifications (Admin/Superuser only)
 */
export async function getAllNotifications(
  limit: number = 100,
  offset: number = 0,
  userId?: string
): Promise<{ notifications: Notification[]; total: number }> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  const params = new URLSearchParams({
    limit: limit.toString(),
    offset: offset.toString(),
  });

  if (userId) params.append('user_id', userId);

  return get(`/notifications/admin/all?${params}`, token);
}

/**
 * Get admin notification stats
 */
export async function getAdminNotificationStats(): Promise<{
  total_notifications: number;
  total_unread: number;
  total_muted: number;
  total_push_subscriptions: number;
}> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  return get('/notifications/admin/stats', token);
}

/**
 * Get system-wide notification configuration
 */
export async function getNotificationConfig(): Promise<{ key: string; value: string; updated_at: string }> {
  return get<{ key: string; value: string; updated_at: string }>('/notifications/config');
}

/**
 * Update system-wide notification configuration (Admin only)
 */
export async function updateNotificationConfig(message: string): Promise<{ key: string; value: string; updated_at: string }> {
  const token = getToken();
  if (!token) throw new Error('Authentication required');

  return patch<{ key: string; value: string; updated_at: string }>(
    '/admin/notifications/config',
    { value: message },
    token
  );
}

// ===== HELPER FUNCTIONS =====

/**
 * Convert VAPID key to Uint8Array
 */
function urlBase64ToUint8Array(base64String: string): Uint8Array {
  const padding = '='.repeat((4 - (base64String.length % 4)) % 4);
  const base64 = (base64String + padding).replace(/-/g, '+').replace(/_/g, '/');

  const rawData = window.atob(base64);
  const outputArray = new Uint8Array(rawData.length);

  for (let i = 0; i < rawData.length; ++i) {
    outputArray[i] = rawData.charCodeAt(i);
  }
  return outputArray;
}

/**
 * Get notification type color
 */
export function getNotificationTypeColor(type: string): string {
  const colors: Record<string, string> = {
    info: 'bg-blue-100 text-blue-800 border-blue-200',
    success: 'bg-green-100 text-green-800 border-green-200',
    warning: 'bg-yellow-100 text-yellow-800 border-yellow-200',
    error: 'bg-red-100 text-red-800 border-red-200',
  };
  return colors[type] || colors.info;
}

/**
 * Get notification type icon
 */
export function getNotificationTypeIcon(type: string): string {
  const icons: Record<string, string> = {
    info: '🔔',
    success: '✅',
    warning: '⚠️',
    error: '❌',
  };
  return icons[type] || icons.info;
}

/**
 * Format notification time
 */
export function formatNotificationTime(dateString: string): string {
  const date = new Date(dateString);
  const now = new Date();
  const diff = now.getTime() - date.getTime();

  const minutes = Math.floor(diff / 60000);
  const hours = Math.floor(diff / 3600000);
  const days = Math.floor(diff / 86400000);

  if (minutes < 1) return 'Just now';
  if (minutes < 60) return `${minutes}m ago`;
  if (hours < 24) return `${hours}h ago`;
  if (days < 7) return `${days}d ago`;

  return date.toLocaleDateString();
}

// ===== WEBSOCKET FOR REAL-TIME NOTIFICATIONS =====

type NotificationCallback = (notification: Notification) => void;
type StatsCallback = (stats: NotificationStats) => void;
type ConnectionCallback = (connected: boolean) => void;

class NotificationWebSocket {
  private ws: WebSocket | null = null;
  private reconnectTimer: NodeJS.Timeout | null = null;
  private pingTimer: NodeJS.Timeout | null = null;
  private notificationCallbacks: Set<NotificationCallback> = new Set();
  private statsCallbacks: Set<StatsCallback> = new Set();
  private connectionCallbacks: Set<ConnectionCallback> = new Set();
  private shouldReconnect: boolean = true;

  /**
   * Connect to WebSocket for real-time notifications
   */
  connect(userId: string): void {
    const token = getToken();
    if (!token || !userId) {
      console.error('Cannot connect to WebSocket: missing token or userId');
      return;
    }

    // Close existing connection
    if (this.ws) {
      this.disconnect();
    }

    try {
      // Get WebSocket URL - includes /api prefix
      const wsBaseUrl = getWebSocketBaseUrl();
      const wsUrl = `${wsBaseUrl}/notifications/ws/${userId}?token=${token}`;
      console.log('Connecting to WebSocket:', wsUrl.replace(token, 'TOKEN_HIDDEN'));
      this.ws = new WebSocket(wsUrl);

      this.ws.onopen = () => {
        console.log('WebSocket connected');
        this.notifyConnectionCallbacks(true);
        this.startPing();
      };

      this.ws.onmessage = (event) => {
        try {
          const data = JSON.parse(event.data);

          if (data.type === 'connected') {
            console.log('WebSocket connection confirmed:', data.message);
          } else if (data.type === 'notification') {
            // New notification received
            this.notifyNotificationCallbacks(data.data);
          } else if (data.type === 'stats_update') {
            // Stats update received
            this.notifyStatsCallbacks(data.data);
          } else if (data.type === 'pong') {
            // Pong received
          }
        } catch (error) {
          console.error('Error parsing WebSocket message:', error);
        }
      };

      this.ws.onerror = (error) => {
        console.error('WebSocket error:', error);
      };

      this.ws.onclose = () => {
        console.log('WebSocket disconnected');
        this.notifyConnectionCallbacks(false);
        this.stopPing();

        // Auto-reconnect after 5 seconds
        if (this.shouldReconnect) {
          this.reconnectTimer = setTimeout(() => {
            console.log('Attempting to reconnect WebSocket...');
            this.connect(userId);
          }, 5000);
        }
      };
    } catch (error) {
      console.error('Failed to create WebSocket connection:', error);
    }
  }

  /**
   * Disconnect from WebSocket
   */
  disconnect(): void {
    this.shouldReconnect = false;

    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }

    this.stopPing();

    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  /**
   * Start sending ping messages to keep connection alive
   */
  private startPing(): void {
    this.stopPing();
    this.pingTimer = setInterval(() => {
      if (this.ws && this.ws.readyState === WebSocket.OPEN) {
        this.ws.send('ping');
      }
    }, 30000); // Send ping every 30 seconds
  }

  /**
   * Stop sending ping messages
   */
  private stopPing(): void {
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
  }

  /**
   * Subscribe to new notifications
   */
  onNotification(callback: NotificationCallback): () => void {
    this.notificationCallbacks.add(callback);
    return () => this.notificationCallbacks.delete(callback);
  }

  /**
   * Subscribe to stats updates
   */
  onStatsUpdate(callback: StatsCallback): () => void {
    this.statsCallbacks.add(callback);
    return () => this.statsCallbacks.delete(callback);
  }

  /**
   * Subscribe to connection status changes
   */
  onConnectionChange(callback: ConnectionCallback): () => void {
    this.connectionCallbacks.add(callback);
    return () => this.connectionCallbacks.delete(callback);
  }

  /**
   * Notify all notification callbacks
   */
  private notifyNotificationCallbacks(notification: Notification): void {
    this.notificationCallbacks.forEach((callback) => {
      try {
        callback(notification);
      } catch (error) {
        console.error('Error in notification callback:', error);
      }
    });
  }

  /**
   * Notify all stats callbacks
   */
  private notifyStatsCallbacks(stats: NotificationStats): void {
    this.statsCallbacks.forEach((callback) => {
      try {
        callback(stats);
      } catch (error) {
        console.error('Error in stats callback:', error);
      }
    });
  }

  /**
   * Notify all connection callbacks
   */
  private notifyConnectionCallbacks(connected: boolean): void {
    this.connectionCallbacks.forEach((callback) => {
      try {
        callback(connected);
      } catch (error) {
        console.error('Error in connection callback:', error);
      }
    });
  }

  /**
   * Check if WebSocket is connected
   */
  isConnected(): boolean {
    return this.ws !== null && this.ws.readyState === WebSocket.OPEN;
  }
}

// Global WebSocket instance
export const notificationWebSocket = new NotificationWebSocket();
